tests/FakeItEasy.Specs/IllegalFluentApiSpecs.cs

```diff
@@ -0,0 +1,189 @@
+            "A.CallTo(() => fake.AnIntMethod()).Throws<System.Exception>().MustHaveHappened()",
+            "does not contain a definition for 'MustHaveHappened'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).Returns(3)",
+            "does not contain a definition for 'Returns'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).SetsOutAndRefParameters('a')",
+            "does not contain a definition for 'SetsOutAndRefParameters'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).WithReturnType<string>()",
+            "does not contain a definition for 'WithReturnType'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).WithNonVoidReturnType()",
+            "does not contain a definition for 'WithNonVoidReturnType'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).Invokes(() => {}).MustHaveHappened()",
+            "does not contain a definition for 'MustHaveHappened'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).Throws<System.Exception>().MustHaveHappened()",
+            "does not contain a definition for 'MustHaveHappened'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).To(7).Returns(3)",
+            "does not contain a definition for 'Returns'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).To(7).SetsOutAndRefParameters('a')",
+            "does not contain a definition for 'SetsOutAndRefParameters'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).To(7).WithReturnType<string>()",
+            "does not contain a definition for 'WithReturnType'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).To(7).WithNonVoidReturnType()",
+            "does not contain a definition for 'WithNonVoidReturnType'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).To(7).Invokes(() => {}).MustHaveHappened()",
+            "does not contain a definition for 'MustHaveHappened'")]
+        [Example(
+            "A.CallToSet(() => fake.AProperty).To(7).Throws<System.Exception>().MustHaveHappened()",
+            "does not contain a definition for 'MustHaveHappened'")]
+        public static void IllegalSyntax(string illegalOperation, string expectedError, string illegalUsage, IEnumerable<Diagnostic> diagnostics)
+        {
+            "Given an illegal usage of a fake"
+                .x(() => illegalUsage = string.Format(CultureInfo.InvariantCulture, FakeUsageTemplate, illegalOperation));
+
+            "When I compile the usage"
+                .x(() => diagnostics = Compile(illegalUsage));
+
+            "Then the compilation fails"
+                .x(() => diagnostics.Should().ContainSingle().Which.GetMessage().Should()
+                    .Contain(expectedError));
+        }
+
+        private static bool CompilationRequiresAssembly(Assembly a)
+        {
+            var assemblyName = a.GetName().Name;
+            return

```

Comment:
I figure this'll have to change. We normally avoid TFMish preprocessing directives, preferring to focus on features, but
1. naming is hard, but we could entertain something like REQUIRES_MSCORLIB_REFERENCE
2. in this case it feels to me like we're really differentiating by framework

I dunno. Like I say, if you prefer something more featurey, I can do. Or if you want more generic, like `NETFRAMEWORK`, `NETCOREAPP`, ... I could pursue. I picked the specific TFM-ish things so we only add what's absolutely required. This'll require us to examine required references if we support another target.

Reply:
Oh, one option that I forgot to mention. It makes the code prettier, but feels gross to me: no `#if`s. The extra references probably aren't going to hurt anything, just make the tests do a little more work.

Reply:
> Oh, one option that I forgot to mention. It makes the code prettier, but feels gross to me: no #ifs.

I like the idea. The code will be much easier to understand, I think.